ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/fastsync/FastSyncCheckpointHeaderManagerTest.java

```diff
@@ -0,0 +1,126 @@
+import tech.pegasys.pantheon.metrics.noop.NoOpMetricsSystem;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class FastSyncCheckpointHeaderManagerTest {
+
+  protected ProtocolSchedule<Void> protocolSchedule;
+  protected EthProtocolManager ethProtocolManager;
+  protected EthContext ethContext;
+  protected ProtocolContext<Void> protocolContext;
+  private SyncState syncState;
+
+  private BlockDataGenerator gen;
+  private BlockchainSetupUtil<Void> localBlockchainSetup;
+  protected MutableBlockchain localBlockchain;
+  private BlockchainSetupUtil<Void> otherBlockchainSetup;
+  protected Blockchain otherBlockchain;
+  private LabelledMetric<OperationTimer> ethTasksTimer;
+  private BlockHeader pivotBlockHeader;
+  private FastSyncCheckpointHeaderManager<Void> checkpointHeaderManager;
+  private RespondingEthPeer peer;
+
+  @Before
+  public void setupTest() {
+    gen = new BlockDataGenerator();
+    localBlockchainSetup = BlockchainSetupUtil.forTesting();
+    localBlockchain = spy(localBlockchainSetup.getBlockchain());
+    otherBlockchainSetup = BlockchainSetupUtil.forTesting();
+    otherBlockchain = otherBlockchainSetup.getBlockchain();
+
+    protocolSchedule = localBlockchainSetup.getProtocolSchedule();
+    protocolContext = localBlockchainSetup.getProtocolContext();
+    ethProtocolManager =
+        EthProtocolManagerTestUtil.create(localBlockchain, localBlockchainSetup.getWorldArchive());
+    ethContext = ethProtocolManager.ethContext();
+    syncState = new SyncState(protocolContext.getBlockchain(), ethContext.getEthPeers());
+
+    ethTasksTimer = NoOpMetricsSystem.NO_OP_LABELLED_TIMER;
+
+    otherBlockchainSetup.importFirstBlocks(30);
+
+    pivotBlockHeader = block(17);
+
+    peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, localBlockchain);
+
+    checkpointHeaderManager =
+        new FastSyncCheckpointHeaderManager<>(
+            SynchronizerConfiguration.builder()
+                .downloaderChainSegmentSize(5)
+                .downloaderHeadersRequestSize(5)
+                .build(),
+            protocolContext,
+            ethContext,
+            syncState,
+            protocolSchedule,
+            ethTasksTimer,
+            pivotBlockHeader);
+  }
+
+  @Test
+  public void shouldNotRequestCheckpointHeadersBeyondThePivotBlock() {
+    final SyncTarget syncTarget = syncState.setSyncTarget(peer.getEthPeer(), block(10));
+    final CompletableFuture<List<BlockHeader>> result =
+        checkpointHeaderManager.pullCheckpointHeaders(syncTarget);
+
+    final Responder responder = RespondingEthPeer.blockchainResponder(otherBlockchain);
+    peer.respondWhile(responder, peer::hasOutstandingRequests);
+
+    assertThat(result).isCompletedWithValue(asList(block(10), block(15), pivotBlockHeader));
+  }
+
+  @Test
+  public void shouldNotDuplicatePivotBlockAsCheckpoint() {
```

Comment:
Might be worth adding a few more edge cases here - when the common ancestor is at the block prior to the pivot, or at the pivot etc.

Reply:
Done.